src/main/java/io/codelex/flightplanner/flights/service/FlightDBService.java

```diff
@@ -0,0 +1,130 @@
+package io.codelex.flightplanner.flights.service;
+
+import io.codelex.flightplanner.flights.repository.AirportRepository;
+import io.codelex.flightplanner.flights.repository.FlightRepository;
+import io.codelex.flightplanner.flights.domain.*;
+import io.codelex.flightplanner.flights.dto.FlightRequest;
+import io.codelex.flightplanner.flights.dto.PageResult;
+import io.codelex.flightplanner.flights.dto.SearchFlightRequest;
+import io.codelex.flightplanner.flights.response.NullException;
+import io.codelex.flightplanner.flights.response.SameFromToException;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import java.text.ParseException;
+
+@Service
+@ConditionalOnProperty(prefix = "myapp", name = "appmode", havingValue = "db")
+public class FlightDBService implements FlightService {
+
+    private FlightRepository flightRepository;
+    private AirportRepository airportRepository;
+
+    public FlightDBService(FlightRepository flightRepository, AirportRepository airportRepository) {
+        this.flightRepository = flightRepository;
+        this.airportRepository = airportRepository;
+    }
+
+    @Override
+    public ResponseEntity<Flight> addFlight(FlightRequest flightRequest) throws ParseException {
+        Flight flight = new Flight(flightRequest);
+        if (flight.checkValidity()) {
+            if (checkUnique(flight)) {
+                flight.setFrom(findOrCreate(flight.getFrom()));
+                flight.setTo(findOrCreate(flight.getTo()));
+                flightRepository.saveAndFlush(flight);
+                return new ResponseEntity<>(flight, HttpStatus.CREATED);
+            }
+            return new ResponseEntity<>(flight, HttpStatus.CONFLICT);
+        } else {
+            return new ResponseEntity<>(flight, HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    @Override
+    public ResponseEntity<Flight> fetchFlight(int id) {
+        Flight flight = flightRepository.findById(id).orElse(null);
+        if (flight == null) {
+            return new ResponseEntity<>(null, HttpStatus.NOT_FOUND);
+        }
+        return new ResponseEntity<>(flight, HttpStatus.OK);
+    }
+
+    @Override
+    public ResponseEntity<Flight> deleteFlight(int id) {
+        if (flightRepository.existsById(id)) {
+            flightRepository.deleteById(id);
+        }
+        return new ResponseEntity<>(null, HttpStatus.OK);
+    }
+
+    @Override
+    public ResponseEntity<Airport[]> searchAirports(String search) {
+        search = search.trim().toLowerCase();
+        for (Airport airport : airportRepository.findAll()) {
```

Comment:
Ar findAll lietošanu DB gadījumā ir jāuzmanās - visbiežāk to nedrīkst lietot.
Ja DB būtu 10 000 ieraksti, tad viņi visu tiktu ielasīti atmiņā un tas var vienā brīdī nokārt sistēmu. Bet ir iespējams uzreiz repo meklēt pēc konkrētiem kritērijiem, vai rakstīt custom meklēšanu, kas notiek DB pusē.
Izlasi šito un arī zemāk tajā pašā rakstā ir manual custom queries:
https://www.baeldung.com/the-persistence-layer-with-spring-data-jpa#1-automatic-custom-queries
Kā ari ir šāds vairants:
https://www.baeldung.com/spring-data-query-by-example#examples

Tas pats iekš checkUnique un findFlights metodēm.

Reply:
Es biju mēģinājis ar automatic custom queries, bet tur neko neatgrieza, jo vienā query es mēģināju meklēt vai nu pēc airport, vai country, vai city un tur vienmēr atgrieza 0 entries.

Reply:
Re kur pāris piemēri no © Zigmāra, bet ņem vērā ka tur vienā ir kļūda, jeb dīvainība, tāpēc pa taisno copy-paste netaisi, bet šādi +/- vajadzētu strādāt:

Airport repo:
```
@Query("SELECT a FROM Airport a WHERE lower(a.airport) LIKE ?1%" +
            "OR lower(a.city) LIKE ?1% OR lower(a.country) LIKE ?1%")
    Set<Airport> searchAirport(String searchString);
```
Flight repo:
```
boolean existsByFromAndToAndCarrierAndDepartureTimeAndArrivalTime(Airport from, Airport to, String carrier, LocalDateTime departureTime, LocalDateTime arrivalTime);

    @Query("Select f from Flight f where f.from.airport = ?1 AND f.to.airport = ?2 " +
            "AND f.departureTime between cast(?3 as timestamp) and cast(?4 as timestamp)")
    List<Flight> searchFlights(String from, String to, LocalDate departureDate, LocalDate departureDatePlusOneDay);
```